byron/ledger/executable-spec/test/Ledger/Delegation/Properties.hs

```diff
@@ -259,18 +271,39 @@ instance HasTrace DBLOCK where
         n <- Gen.integral (Range.linear 0 13)
         pure $! Set.fromAscList $ mkVKeyGenesis <$> [0 .. n]
 
-  sigGen _ (env, st) =
-    DBlock <$> nextSlotGen <*> sigGen @DELEG env st
+  sigGen (Just InvalidDelegationCertificate) _ (env, _st) =
+    DBlock <$> nextSlotGen env <*> Gen.list (Range.constant 0 10) randomDCertGen
     where
-      -- We want the resulting trace to include a large number of epoch
-      -- changes, so we generate an epoch change with higher frequency.
-      nextSlotGen =
-        incSlot <$> Gen.frequency
-                      [ (1, Gen.integral (Range.constant 1 10))
-                      , (2, pure $! slotsPerEpoch (_dSEnvK env))
-                      ]
+      -- | Generate a random delegation certificate, which has a high probability of failing since
+      -- we do not consider the current delegation state. So for instance, we could generate a
+      -- delegation certificate for a genesis key that already delegated in this epoch.
+      randomDCertGen :: Gen DCert
+      randomDCertGen = do
+        (vkg, vk, e) <- (,,) <$> vkgGen' <*> vkGen' <*> epochGen'
+        pure $! mkDCert vkg (signWithGenesisKey vkg (vk, e)) vk e
         where
-          incSlot c = (env ^.slot) `addSlot` SlotCount c
+          vkgGen' = Gen.element $ Set.toList allowed
+          allowed = _dSEnvAllowedDelegators env
+          vkGen' = Gen.element $ VKey . Owner <$> [0 .. (2 * fromIntegral (length allowed))]
+          epochGen' =  Epoch
+                    .  fromIntegral -- We don't care about underflow. We want to generate large epochs anyway.
+                    .  (fromIntegral n +)
+                   <$> Gen.integral (Range.constant (-2 :: Int) 2)
+            where Epoch n = _dSEnvEpoch env
+  sigGen _ _ (env, st) =
+    DBlock <$> nextSlotGen env <*> sigGen @DELEG Nothing env st
+
+
+-- | Generate a next slot. We want the resulting trace to include a large number of epoch changes,
+-- so we generate an epoch change with higher frequency.
+nextSlotGen :: DSEnv -> Gen Slot
+nextSlotGen env =
+  incSlot <$> Gen.frequency
+                [ (1, Gen.integral (Range.constant 1 10))
+                , (2, pure $! slotsPerEpoch (_dSEnvK env))
+                ]
+  where
+    incSlot c = (env ^.slot) `addSlot` SlotCount c
```

Comment:
If I understand correctly, this will either (1) bump the slot (likely within the same epoch) by a random number in range 1-10, or (2) add enough slots to move us one full epoch forward, but at the same slot. Might it be useful or interesting to add a random number of slots to case (2)?

Something like `(2, (+ (slotsPerEpoch (_dSEnvK env))) <$> Gen.integral (Range.constant 1 10))`? Just wondering aloud.

Reply:
Yeah, definitely! If we do we need to examine how this would impact coverage (but I don't expect to see any difference, since we should see roughly the same amount of epoch changes....).